service/history/historyEngine.go

```diff
@@ -637,6 +637,11 @@ func (e *historyEngineImpl) getMutableState(
 		BranchToken:            msBuilder.GetCurrentBranch(),
 	}
 
+	if !msBuilder.IsWorkflowExecutionRunning() {
```

Comment:
3+DC will have additional zombie state, perhaps, you should check whether the closeStatus is not 'none'